src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs

```diff
@@ -288,20 +300,199 @@ private static ProjectItemGroupElement CreateItemGroup(Project project, string f
+        }
+
+        /// <summary>
+        /// Prepares the dictionary that maps frameworks to packages top-level
+        /// and transitive.
+        /// </summary>
+        /// <param name="project"></param>
+        /// <param name="userInputFrameworks">A list of framework names</param>
+        /// <param name="assetsFile">Assets file for all targets and libraries</param>
+        /// <param name="transitive">Include transitive packages in the result</param>
+        /// <returns></returns>
+        internal IEnumerable<FrameworkPackages> GetResolvedVersions(
+            Project project, IEnumerable<string> userInputFrameworks,LockFile assetsFile, bool transitive)
+        {
+            var resultPackages = new List<FrameworkPackages>();
+            var parsedUserFrameworks = userInputFrameworks.Select(f => NuGetFramework.Parse(f));
+
+            IEnumerable<TargetFrameworkInformation> requestedTargetFrameworks;
+
+            if (!parsedUserFrameworks.Any())
+            {
+                requestedTargetFrameworks = assetsFile.PackageSpec.TargetFrameworks;
+            }
+            else
+            {
+                requestedTargetFrameworks = assetsFile.PackageSpec.TargetFrameworks.Where(tfm => parsedUserFrameworks.Contains(tfm.FrameworkName));
+            }
+
+            var requestedTargets = requestedTargetFrameworks.Select(tfm => assetsFile.GetTarget(tfm.FrameworkName, null));
+
+            foreach (var target in requestedTargets)
+            {
+                // Find the tfminformation corresponding to the target to
+                // get the top-level dependencies
+                var tfmInformation = requestedTargetFrameworks.Where(tfm => tfm.FrameworkName.Equals(target.TargetFramework));
+
+                //The packages for the framework that were retrieved with GetRequestedVersions
+                var frameworkDependencies = tfmInformation.First().Dependencies;
+
+                var projPackages = GetPackageReferencesPerFramework(project, "", tfmInformation.ToString());
+
+                var topLevelPackages = new List<InstalledPackageReference>();
+                var transitivePackages = new List<InstalledPackageReference>();
+
+                foreach (var library in target.Libraries)
+                {
+                    var matchingPackages = frameworkDependencies.Where(d =>
+                        d.Name.Equals(library.Name, StringComparison.OrdinalIgnoreCase));
+
+                    //In case we found a matching package in requestedVersions, the package will be
+                    //top level. If not, then it is transitive, and include-transitive must be used
+                    //to add the package
+                    if (matchingPackages.Any() || transitive)
+                    {
+                        
+                        var resolvedVersion = library.Version.ToString();
+
+                        if (matchingPackages.Any())
```

Comment:
you can simplified this and the above condition... have two separate if conditions